src/main/java/org/opensearch/knn/index/query/filtered/FilteredIdsKNNByteIterator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.filtered;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.BitSetIterator;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.knn.index.SpaceType;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+/**
+ * Inspired by DiversifyingChildrenFloatKnnVectorQuery in lucene
+ * https://github.com/apache/lucene/blob/7b8aece125aabff2823626d5b939abf4747f63a7/lucene/join/src/java/org/apache/lucene/search/join/DiversifyingChildrenFloatKnnVectorQuery.java#L162
+ *
+ * The class is used in KNNWeight to score filtered KNN field by iterating filterIdsArray.
+ */
+public class FilteredIdsKNNByteIterator implements KNNIterator {
+    // Array of doc ids to iterate
+    protected final BitSet filterIdsBitSet;
+    protected final BitSetIterator bitSetIterator;
+    protected final byte[] queryVector;
+    protected final BinaryDocValues binaryDocValues;
+    protected final SpaceType spaceType;
+    protected float currentScore = Float.NEGATIVE_INFINITY;
+    protected int docId;
+
+    public FilteredIdsKNNByteIterator(
+        final BitSet filterIdsBitSet,
+        final byte[] queryVector,
+        final BinaryDocValues binaryDocValues,
+        final SpaceType spaceType
+    ) {
+        this.filterIdsBitSet = filterIdsBitSet;
+        this.bitSetIterator = new BitSetIterator(filterIdsBitSet, filterIdsBitSet.length());
+        this.queryVector = queryVector;
+        this.binaryDocValues = binaryDocValues;
+        this.spaceType = spaceType;
+        this.docId = bitSetIterator.nextDoc();
+    }
+
+    /**
+     * Advance to the next doc and update score value with score of the next doc.
+     * DocIdSetIterator.NO_MORE_DOCS is returned when there is no more docs
+     *
+     * @return next doc id
+     */
+    @Override
+    public int nextDoc() throws IOException {
+
+        if (docId == DocIdSetIterator.NO_MORE_DOCS) {
+            return DocIdSetIterator.NO_MORE_DOCS;
+        }
+        int doc = binaryDocValues.advance(docId);
+        currentScore = computeScore();
+        docId = bitSetIterator.nextDoc();
+        return doc;
+    }
+
+    @Override
+    public float score() {
+        return currentScore;
+    }
+
+    protected float computeScore() throws IOException {
+        final BytesRef value = binaryDocValues.binaryValue();
+        final ByteArrayInputStream byteStream = new ByteArrayInputStream(value.bytes, value.offset, value.length);
+        final byte[] vector = byteStream.readAllBytes();
```

Comment:
This should be behind the VectorSerializer interfaces otherwise there is no common way to read the binary vectors/float vectors.